### PR TITLE
feat(pruner): Add config based pruner

### PR DESCRIPTION
**Pull request by zechmeister**
This introduces a pruning service that _prunes_/filters/reduces unnecessary properties from userData (`Context`) by combining the flow configuration and flowPages from strapi.

It does so in three steps:
1. use userData and the flowConfiguration and follow the Paths a user has taken through the flow to collect all steps a user has seen (`function getPaths(...): StepPath[]`). The steps are organised into lists of steps (`Paths`) because subflows and arrays need to be treated differently.
2. get the associated formFields from strapi for every step from 1.
3. pick the data from userData that has representations in the formFields from 2.

## todo:
- [x] fetch the flowPages only once (relevant for performance on staging)
- [ ] ~~e2e tests~~ integration tests
- [ ] move `getSteps()` to FlowController

### CODE REVIEW

File: app/services/flow/pruner.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  FlowController,
+  buildFlowController,
+} from "./server/buildFlowController";
+import { ArrayConfig } from "../array";
+import { fetchFlowPage } from "../cms/index.server";
+
+type Path = { steps: string[]; arrayIndex?: number };
+type FormField = { name: string; arrayIndex?: number };
+
+export async function prune(
```
Comment (frederike-ramin) [inline app/services/flow/pruner.ts:16]: Can we call this `pruneIrrelevantData` or `filterRelevantData` or similar? And then probably also rename the file accordingly?

File: app/services/flow/pruner.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        steps: getSteps(
+          buildFlowController({
+            config: flowController.getConfig(),
+            data: addPageDataToUserData(userData, {
+              arrayIndexes: [index],
+            }),
+            guards: flowController.getGuards(),
+          }),
+          getSubFlowInitialStep(array.config),
+        ).filter(stepBelongsToArray(array.config)),
```
Comment (frederike-ramin) [inline app/services/flow/pruner.ts:112]: Why do we have to filter for relevant array steps again? Do we link from the array subflow back to the uebersicht page on SUBMIT? Maybe we should rethink that then?

File: app/services/flow/pruner.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  );
+}
+
+export function getPropsToKeep(formFields: FormField[]): string[] {
+  return formFields.map((formField) =>
+    formField.name.replace("#", `[${formField.arrayIndex}].`),
+  );
+}
+
+function getSteps(
```
Comment (frederike-ramin) [inline app/services/flow/pruner.ts:69]: I'm wondering if this should be a function on the `flowController`?

File: app/services/flow/pruner.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+): Promise<FormField[]> {
+  const steps = paths.map((path) => path.steps).flat();
+
+  // to be replaced with new fetching
+  const formFields: { [stepId: string]: string[] } = Object.fromEntries(
+    await Promise.all(
+      steps.map(async (step) => [
+        step,
+        (await fetchFlowPage("form-flow-pages", flowId, step)).form.flatMap(
+          (formItem) => formItem.name,
```
Comment (frederike-ramin) [inline app/services/flow/pruner.ts:50]: Can we call this `formComponent`?

File: app/services/flow/pruner.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+export async function getFormFields(
+  paths: Path[],
+  flowId: FlowId,
+): Promise<FormField[]> {
+  const steps = paths.map((path) => path.steps).flat();
+
+  // to be replaced with new fetching
+  const formFields: { [stepId: string]: string[] } = Object.fromEntries(
```
Comment (frederike-ramin) [inline app/services/flow/pruner.ts:45]: actually pretty crazy how neat that function is while it does so much! 🚀

File: app/services/flow/pruner.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        ),
+      ]),
+    ),
+  );
+
+  return paths.flatMap(({ steps, arrayIndex }) =>
+    steps.flatMap((step) =>
+      formFields[step].map((name) => ({ name, arrayIndex })),
+    ),
+  );
```
Comment (frederike-ramin) [inline app/services/flow/pruner.ts:60]: Okay, this I don't get what it does. :D

**Review by frederike-ramin — COMMENTED** [review]
I love it. Absolutely do. I left quite a few comments, but they're mostly for better understanding, not challenging the implementation in its whole. 
It's a very nice solution to a very nasty problem. :)

File: app/services/flow/__test__/pruner.test.ts [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+          steps: [
+            "finanzielle-angaben/eigentum-zusammenfassung/bankkonten/daten",
+          ],
+          arrayIndex: 0,
+        },
+      ]);
+    });
+
+    // todo check for arrays in place
+    it("includes multiple paths for subflows", () => {
```
[the remark this answers is not in the document]
Comment (zechmeister) [reply app/services/flow/__test__/pruner.test.ts:91]: we can—by moving it to the flowController and using the test setup there!

File: app/services/flow/__test__/pruner.test.ts [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+        "step1",
+      );
+      expect(fetchFlowPageMock).toHaveBeenCalledWith(
+        "form-flow-pages",
+        "/beratungshilfe/antrag",
+        "step2",
+      );
+    });
+
+    it("returns FornamePaths", async () => {
```
[the remark this answers is not in the document]
Comment (zechmeister) [reply app/services/flow/__test__/pruner.test.ts]: check out [this commit](https://github.com/digitalservicebund/a2j-rechtsantragstelle/pull/1074/commits/8b0d1a5da5239776b4df2f0cf94e836a42a2ad1f) for multi formFields testing

**Comment by frederike-ramin** [comment]
Only real steps left to do from my point of view:
1. Fix e2e tests
2. add Feature Flag
3. Let product/design test it
4. Disable feature flag

I think all the other changes + refactorings we can do after merging. wdyt?

**Pull request merged**